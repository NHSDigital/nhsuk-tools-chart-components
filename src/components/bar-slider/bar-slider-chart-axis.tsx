import { BarSliderChartModel } from '@models/bar-slider-chart.model';
import React, { JSX } from 'react';

type BarSliderGraphAxisMarkerProps = {
  chartName: string;
  markerText: string;
  offsetPercent: number;
};

type BarSliderGraphAxisProps = {
  model: BarSliderChartModel;
};

const markerAxisLabelClassname = (
  name: string,
  offsetPercent: number
): string =>
  offsetPercent > 0
    ? `${name}-axis-marker-label`
    : `${name}-start-axis-marker-label`;

const markerAxisClassname = (name: string, offsetPercent: number): string =>
  offsetPercent > 0 ? `${name}-axis-marker` : `${name}-start-axis-marker`;

function AxisMarker({
  chartName,
  markerText,
  offsetPercent,
}: Readonly<BarSliderGraphAxisMarkerProps>): JSX.Element {
  return (
    <div
      className={`${markerAxisClassname(chartName, offsetPercent)}`}
      style={{ left: `${offsetPercent}%` }}>
      <div
        className={`nhsuk-font-size-16 nhsuk-u-secondary-text-color ${markerAxisLabelClassname(
          chartName,
          offsetPercent
        )}`}>
        {markerText}
      </div>
    </div>
  );
}

export function BarSliderChartAxis({
  model,
}: Readonly<BarSliderGraphAxisProps>): JSX.Element {
  return (
    <>
      {model
        .getAxisTicks()
        .map(
          ({ label, value, isVisible }) =>
            isVisible && (
              <AxisMarker
                chartName={model.name}
                markerText={label}
                offsetPercent={value}
                key={label}
              />
            )
        )}
    </>
  );
}
