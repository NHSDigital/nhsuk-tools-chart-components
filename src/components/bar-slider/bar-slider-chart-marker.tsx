import React, { JSX } from 'react';
import { ArrowMarkerIcon } from '@components/markers/arrow-marker-icon';
import { BarSliderChartModel } from '@models/bar-slider-chart.model';
import './styles/bar-slider-chart-marker.module.scss';

const markerAlighTextClassname = (name: string, offset: number): string => {
  if (offset <= 20) {
    return `${name}-label-align-left`;
  }
  if (offset <= 80) {
    return `${name}-label-align-center`;
  }
  return `${name}-label-align-right`;
};

type BarSliderChartMarkerTextProps = Readonly<{
  chartName: string;
  value: number;
  offset: number;
  markerText: string;
}>;

function BarSliderChartMarkerText({
  chartName,
  value,
  offset,
  markerText,
}: Readonly<BarSliderChartMarkerTextProps>): JSX.Element {
  return (
    <div className={`${chartName}-label-marker`}>
      <div
        className={`nhsuk-font-size-16 ${markerAlighTextClassname(
          chartName,
          offset
        )}`}>
        {markerText}
        <span className={`${chartName}-value-marker`}>{value}</span>
      </div>
    </div>
  );
}
export function BarSliderChartMarker({
  value,
  model,
  markerText,
}: Readonly<{
  value: number;
  model: BarSliderChartModel;
  markerText: string;
}>): JSX.Element {
  const offset = model.calculatePosition(value);

  const roundingPrecision = 1;
  const multiplier = 10 ** roundingPrecision;
  const roundedOffset = Math.round(offset * multiplier) / multiplier;
  return (
    <>
      <div
        className={`${model.name}-arrow-marker`}
        style={{ left: `${roundedOffset}%` }}>
        <BarSliderChartMarkerText
          chartName={model.name}
          value={value}
          markerText={markerText}
          offset={roundedOffset}
        />
        <ArrowMarkerIcon />
      </div>
      <div
        className={`${model.name}-value-axis-marker`}
        style={{ left: `${roundedOffset}%` }}
      />
    </>
  );
}
