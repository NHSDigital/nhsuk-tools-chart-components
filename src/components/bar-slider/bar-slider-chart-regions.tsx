import { BarSliderChartModel } from '@models/bar-slider-chart.model';
import React, { JSX } from 'react';

function RegionForClassification({
  value,
  styleClass,
}: Readonly<{
  value: number;
  styleClass: string;
}>): JSX.Element {
  return <div className={styleClass} style={{ width: `${value}%` }} />;
}

export function BarSliderChartRegions({
  model,
}: Readonly<{
  model: BarSliderChartModel;
}>): JSX.Element {
  return (
    <>
      {model.getChartRegions().map(({ value, styleClass }) => (
        <RegionForClassification
          value={value}
          styleClass={styleClass}
          key={value}
        />
      ))}
    </>
  );
}
