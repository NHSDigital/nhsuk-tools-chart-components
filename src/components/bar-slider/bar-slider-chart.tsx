import React, { JSX } from 'react';
import { BarSliderChartModel } from '@models/bar-slider-chart.model';
import { BarSliderConfig } from '@custom-types/bar-slider.types';
import { BarSliderChartAxis } from './bar-slider-chart-axis';
import { BarSliderChartMarker } from './bar-slider-chart-marker';
import { BarSliderChartRegions } from './bar-slider-chart-regions';
import './styles/bar-slider-chart.module.scss';

const CHART_WIDTH = 100;

/**
 * The properties for @BarSliderChart component
 *
 * @param {BarSliderConfig} BarSliderConfig -  The bar slider chart configuration
 * @param {number} value - The value to render in the chart
 * @param {string} markerText - Marker text for the value
 */
export type BarSliderChartProps = {
  config: BarSliderConfig;
  value: number;
  markerText: string;
};

/**
 * A component to display the bar slider chart
 *
 * @typedef {object} BarSliderChartProps
 * @property {BarSliderChartProps} config -  The bar slider chart configuration
 * @property {number} value - The value to render in the chart
 * @property {string} markerText - Marker text for the value
 *
 * @param {BarSliderChartProps} props
 * @returns {JSX.Element}
 */
export function BarSliderChart({
  config,
  value,
  markerText,
}: Readonly<BarSliderChartProps>): JSX.Element {
  const chartModel = new BarSliderChartModel(config, CHART_WIDTH);

  return (
    <div
      className={`${config.name}-graph nhsuk-u-margin-bottom-9`}
      aria-hidden="true">
      <BarSliderChartAxis model={chartModel} />
      <BarSliderChartMarker
        value={value}
        markerText={markerText}
        model={chartModel}
      />
      <BarSliderChartRegions model={chartModel} />
    </div>
  );
}
