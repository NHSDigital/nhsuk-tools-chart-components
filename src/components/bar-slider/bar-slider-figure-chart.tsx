import React, { JSX } from 'react';
import { ChartLegend } from '@components/chart-legend';
import { BarSliderConfig } from '@custom-types/bar-slider.types';
import { BarSliderChart } from './bar-slider-chart';

/**
 * The properties for @BarSliderChartFigureProps component
 *
 * @param {BarSliderConfig} config -  The bar slider chart configuration
 * @param {number} value - The value to render in the chart
 * @param {string} ariaLabel - The aria label for the chart figure
 * @param {string} markerText - The marker text for value
 */
export type BarSliderChartFigureProps = {
  config: BarSliderConfig;
  value: number;
  ariaLabel: string;
  markerText: string;
};

/**
 * A component to display the bar slider chart figure
 *
 * @typedef {object} BarSliderChartFigureProps
 * @param {BarSliderConfig} config -  The bar slider chart configuration
 * @param {number} value - The value to render in the chart
 * @param {string} ariaLabel - The aria label for the chart figure
 * @param {string} markerText - The marker text for value
 * @param {BarSliderChartFigureProps} props
 *
 * @returns {JSX.Element}
 */
export function BarSliderChartFigure({
  config,
  value,
  ariaLabel,
  markerText,
}: Readonly<BarSliderChartFigureProps>): JSX.Element {
  const barSliderLegendKeys = Object.values(config.segments)
    .filter((s) => s.legend != null)
    .map((segment) => ({
      text: segment.legend!,
      className: `${config.name}-legend-key-${segment.styleSuffix}`,
      wrapperClassName: `${config.name}-legend-marker`,
    }));

  return (
    <figure className="nhsuk-u-margin-0">
      <figcaption className="nhsuk-u-visually-hidden">{ariaLabel}</figcaption>
      <BarSliderChart config={config} value={value} markerText={markerText} />
      <ChartLegend legendKeys={barSliderLegendKeys} borderWrapper={false} />
    </figure>
  );
}
