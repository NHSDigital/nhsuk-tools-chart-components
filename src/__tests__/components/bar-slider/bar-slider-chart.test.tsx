import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BarSliderChart } from '@components/bar-slider/bar-slider-chart';
import { BarSliderConfig } from '@custom-types/bar-slider.types';
import { mockBarSliderConfigWithConstraintsExample } from '@tests/__helpers__/mocks/bar-slider-results';

const markerTest = 'Your Result:';

const mockConfig: BarSliderConfig = {
  name: 'nhsuk-bar-slider',
  axis: {
    ticks: [
      { value: 0, isVisible: false },
      { value: 10, isVisible: true },
      { value: 40, isVisible: true },
      { value: 60, isVisible: true },
      { value: 90, isVisible: true },
      { value: 100, isVisible: false },
    ],
  },
  segments: [
    {
      styleSuffix: 'mock-style-1',
      weight: 10,
    },
    {
      styleSuffix: 'mock-style-2',
      weight: 30,
    },
    {
      styleSuffix: 'mock-style-3',
      weight: 20,
    },
    {
      styleSuffix: 'mock-style-4',
      weight: 30,
    },
    {
      styleSuffix: 'mock-style-5',
      weight: 10,
    },
  ],
};

function BarSliderTestApp({
  config = mockConfig,
  value = 15.9,
  markerText = markerTest,
}: Readonly<{
  config?: BarSliderConfig;
  value?: number;
  markerText?: string;
}>): JSX.Element {
  return (
    <BarSliderChart config={config} value={value} markerText={markerText} />
  );
}

describe('The Bar slider chart component', () => {
  it('Renders with the expected content', () => {
    const { container } = render(<BarSliderTestApp />);
    expect(container).toMatchSnapshot();
  });

  it('Renders the markerText', () => {
    const markerText = 'Mocked Result:';
    const { container } = render(
      <BarSliderTestApp markerText={markerText} value={85} />
    );

    const labelMarkerEl = container?.querySelector(
      '.nhsuk-bar-slider-label-marker'
    );
    expect(labelMarkerEl).not.toBeNull();
    expect(labelMarkerEl?.innerHTML).toContain(markerText);

    expect(screen.getByText(markerText).classList[1]).toContain(
      'nhsuk-bar-slider-label-align-right'
    );
  });

  it.each`
    value  | labelAlignment
    ${1}   | ${'left'}
    ${10}  | ${'left'}
    ${20}  | ${'left'}
    ${21}  | ${'center'}
    ${50}  | ${'center'}
    ${80}  | ${'center'}
    ${81}  | ${'right'}
    ${90}  | ${'right'}
    ${100} | ${'right'}
  `(
    'with chart $value expect the label to be positioned in $labelAlignment',
    async ({
      value,
      labelAlignment,
    }: {
      value: number;
      labelAlignment: string;
    }) => {
      const markerText = 'Mocked Result:';
      const { container } = render(
        <BarSliderTestApp markerText={markerText} value={value} />
      );

      const labelMarkerEl = container?.querySelector(
        '.nhsuk-bar-slider-label-marker'
      );
      expect(labelMarkerEl).not.toBeNull();
      expect(labelMarkerEl?.innerHTML).toContain(markerText);

      expect(screen.getByText(markerText).classList[1]).toContain(
        `nhsuk-bar-slider-label-align-${labelAlignment}`
      );
    }
  );

  it.each`
    value  | expectedArrowPosition
    ${1}   | ${1}
    ${10}  | ${10}
    ${50}  | ${50}
    ${60}  | ${60}
    ${90}  | ${90}
    ${100} | ${100}
  `(
    'expect the chart arrow marker axis to be positioned relative to value with $expectedArrowPosition',
    async ({
      value,
      expectedArrowPosition,
    }: {
      value: number;
      expectedArrowPosition: number;
    }) => {
      const { container } = render(<BarSliderTestApp value={value} />);

      const arrowMarkerEl = container?.querySelector(
        '.nhsuk-bar-slider-arrow-marker'
      ) as HTMLElement;
      expect(arrowMarkerEl).not.toBeNull();
      expect(arrowMarkerEl.style.left).toBe(`${expectedArrowPosition}%`);
    }
  );

  it('Renders the chart regions', () => {
    const { container } = render(<BarSliderTestApp />);

    const expectedWeights = [10, 40, 60, 90, 100];
    mockConfig.segments.forEach((tick, id) => {
      const regionEl = container?.querySelector(
        `.nhsuk-bar-slider-mock-style-${id + 1}`
      ) as HTMLElement;
      expect(regionEl).not.toBeNull();
      expect(regionEl.style.width).toBe(`${expectedWeights[id]}%`);
    });
  });

  it('Renders the chart axis ticks', () => {
    const { container } = render(<BarSliderTestApp />);

    const expectedWeights = [10, 40, 60, 90];
    const axisTickElements = container?.querySelectorAll(
      `.nhsuk-bar-slider-axis-marker`
    );

    expect(axisTickElements.length).toBe(4);
    axisTickElements.forEach((axisEl, id) => {
      expect((axisEl as HTMLElement).style.left).toBe(
        `${expectedWeights[id]}%`
      );
      const axisLabelElement = axisEl?.querySelector(
        '.nhsuk-bar-slider-axis-marker-label'
      );
      expect(axisLabelElement!.innerHTML).toBe(`${expectedWeights[id]}`);
    });
  });

  it.each`
    value
    ${1}
    ${11.23}
    ${45.2}
    ${67.1}
    ${100}
  `(
    'expect the chart to contain the label value as $value',
    async ({ value }: { value: number }) => {
      const { container } = render(<BarSliderTestApp value={value} />);

      const ratioMarkerEl = container?.querySelector(
        '.nhsuk-bar-slider-value-marker'
      );
      expect(ratioMarkerEl).not.toBeNull();
      expect(ratioMarkerEl?.innerHTML).toContain(value.toString());
    }
  );

  it.each`
    value   | expectedAxisPosition
    ${0.1}  | ${0}
    ${0.4}  | ${0}
    ${0.5}  | ${38}
    ${0.55} | ${57}
    ${0.6}  | ${76}
    ${0.64} | ${95.2}
    ${0.65} | ${98}
    ${1}    | ${98}
  `(
    'expect the chart arrow market to be positioned in $expectedAxisPosition and clamped if $value exceeds max constraints ',
    async ({
      value,
      expectedAxisPosition,
    }: {
      value: number;
      expectedAxisPosition: number;
    }) => {
      const { container } = render(
        <BarSliderTestApp
          config={mockBarSliderConfigWithConstraintsExample}
          value={value}
        />
      );

      const arrowMarkerEl = container?.querySelector(
        '.nhsuk-bar-slider-arrow-marker'
      ) as HTMLElement;
      expect(arrowMarkerEl).not.toBeNull();
      expect(arrowMarkerEl.style.left).toBe(`${expectedAxisPosition}%`);
    }
  );
});
