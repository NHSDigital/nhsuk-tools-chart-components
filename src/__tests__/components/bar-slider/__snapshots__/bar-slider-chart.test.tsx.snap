// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Bar slider chart component Renders with the expected content 1`] = `
<div>
  <div
    aria-hidden="true"
    class="nhsuk-bar-slider-graph nhsuk-u-margin-bottom-9"
  >
    <div
      class="nhsuk-bar-slider-axis-marker"
      style="left: 10%;"
    >
      <div
        class="nhsuk-font-size-16 nhsuk-u-secondary-text-color nhsuk-bar-slider-axis-marker-label"
      >
        10
      </div>
    </div>
    <div
      class="nhsuk-bar-slider-axis-marker"
      style="left: 40%;"
    >
      <div
        class="nhsuk-font-size-16 nhsuk-u-secondary-text-color nhsuk-bar-slider-axis-marker-label"
      >
        40
      </div>
    </div>
    <div
      class="nhsuk-bar-slider-axis-marker"
      style="left: 60%;"
    >
      <div
        class="nhsuk-font-size-16 nhsuk-u-secondary-text-color nhsuk-bar-slider-axis-marker-label"
      >
        60
      </div>
    </div>
    <div
      class="nhsuk-bar-slider-axis-marker"
      style="left: 90%;"
    >
      <div
        class="nhsuk-font-size-16 nhsuk-u-secondary-text-color nhsuk-bar-slider-axis-marker-label"
      >
        90
      </div>
    </div>
    <div
      class="nhsuk-bar-slider-arrow-marker"
      style="left: 15.9%;"
    >
      <div
        class="nhsuk-bar-slider-label-marker"
      >
        <div
          class="nhsuk-font-size-16 nhsuk-bar-slider-label-align-left"
        >
          Your Result:
          <span
            class="nhsuk-bar-slider-value-marker"
          >
            15.9
          </span>
        </div>
      </div>
      <svg
        fill="none"
        viewBox="0 0 12 12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9L0.803849 -9.78799e-07L11.1962 -7.02746e-08L6 9Z"
          fill="black"
        />
      </svg>
    </div>
    <div
      class="nhsuk-bar-slider-value-axis-marker"
      style="left: 15.9%;"
    />
    <div
      class="nhsuk-bar-slider-mock-style-1"
      style="width: 10%;"
    />
    <div
      class="nhsuk-bar-slider-mock-style-2"
      style="width: 40%;"
    />
    <div
      class="nhsuk-bar-slider-mock-style-3"
      style="width: 60%;"
    />
    <div
      class="nhsuk-bar-slider-mock-style-4"
      style="width: 90%;"
    />
    <div
      class="nhsuk-bar-slider-mock-style-5"
      style="width: 100%;"
    />
  </div>
</div>
`;
